dawn/src/dawn/Validator/TypeChecker.cpp

```diff
@@ -85,88 +99,304 @@ void TypeChecker::TypeCheckerImpl::visit(
     }
   }
 
-  DAWN_ASSERT(nameToLocationType_.count(fieldName));
-  curType_ = nameToLocationType_.at(fieldName);
-} // namespace dawn
+  DAWN_ASSERT(nameToDimensions_.count(fieldName));
+  curDimensions_ = nameToDimensions_.at(fieldName);
+}
+
+void TypeChecker::TypeCheckerImpl::checkBinaryOpTriangular(const sir::FieldDimensions& left,
+                                                           const sir::FieldDimensions& right) {
+  const auto& triangularDimLeft = getTriangularDim(left);
+  const auto& triangularDimRight = getTriangularDim(right);
+
+  // Case 0: Both operands are sparse. Their dense + sparse parts must match.
```

Comment:
We had a design document which shows all the cases, no? Please add it to the wiki as reference.